File: packages/agentenkit/test/thread-start.test.ts
import { describe, expect, it } from 'bun:test';
import { simulateReadableStream } from 'ai';
import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import { MockLanguageModelV1 } from 'ai/test';
import { setupAgentCore } from '../src/runtime.js';
import { MemoryAdminStore } from '../src/admin/memory.js';
import { SqliteAdminStore } from '../src/admin/sqlite.js';
import { openSqlite } from '../src/adapters/sqlite.js';
import { MemoryBus, MemoryKv, MemoryQueue, MemoryStorage } from '../src/adapters/memory.js';
import { resolveConfig, type AgentConfig } from '../src/core/types.js';
import type { AdminStore } from '../src/ports/admin.js';
import type { RuntimeOptions } from '../src/ports/runtime.js';

const model = () =>
  new MockLanguageModelV1({
    provider: 'mock', modelId: 'mock',
    doStream: async () => ({
      stream: simulateReadableStream({
        chunks: [
          { type: 'text-delta', textDelta: 'ok' },
          { type: 'finish', finishReason: 'stop', usage: { promptTokens: 10, completionTokens: 5 } },
        ] as LanguageModelV1StreamPart[],
      }),
      rawCall: { rawPrompt: null, rawSettings: {} },
    }),
  });

async function makeRuntime(config: Partial<AgentConfig> = {}, admin: AdminStore = new MemoryAdminStore()) {
  const queue = new MemoryQueue();
  const deps: RuntimeOptions = {
    storage: new MemoryStorage(), admin, bus: new MemoryBus(), queue, kv: new MemoryKv(),
    resolveModel: () => ({ instance: () => model(), contextWindow: 128_000 }),
    config: resolveConfig({ providerOptions: { openai: { tier: 'flex' } }, ...config }),
  };
  return { runtime: await setupAgentCore(deps), queue, admin };
}

describe('what started a thread (§2.9)', () => {
  it('is recorded on the thread on first sight, with every dispatch parameter', async () => {
    const { runtime, queue } = await makeRuntime();
    const chat = runtime.createStreamTextAgent({
      name: 'chat', model: 'gpt-4o', providerOptions: { anthropic: { effort: 'high' } },
    });
    const ran = await chat.run({
      prompt: 'send the quarterly report', tokenBudget: 5_000,
      state: { orgId: 'acme' }, providerOptions: { openai: { tier: 'priority' } },
    });
    const [thread] = await runtime.admin.listThreads();
    expect(thread!.startedWith).toMatchObject({
      runId: ran.runId, agent: 'chat', model: 'gpt-4o',
      prompt: 'send the quarterly report', tokenBudget: 5_000, state: { orgId: 'acme' },
      // The three levels merged, the input winning per namespace (§3.1)
      providerOptions: { openai: { tier: 'priority' }, anthropic: { effort: 'high' } },
    });
    expect(thread!.startedWith!.at).toBeInstanceOf(Date);
    expect(thread!.prompt).toBe('send the quarterly report');
    // The run record carries the same provider options next to its prompt
    const run = (await runtime.admin.getRun(ran.runId!))!.run;
    expect(run.providerOptions).toEqual({ openai: { tier: 'priority' }, anthropic: { effort: 'high' } });

    // A second run on the thread never overwrites what started it
    await runtime.worker.handleJob(queue.items[0]!);
    const again = await chat.run({ threadId: ran.threadId, prompt: 'and again' });
    const [after] = await runtime.admin.listThreads();
    expect(after!.startedWith!.runId).toBe(ran.runId!);
    expect(after!.startedWith!.prompt).toBe('send the quarterly report');
    expect(again.runId).not.toBe(ran.runId);
  });

  it('keeps only the identity when recordPayloads is off', async () => {
    const { runtime } = await makeRuntime({ recordPayloads: false });
    const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
    const ran = await chat.run({ prompt: 'private', state: { orgId: 'acme' }, providerOptions: { openai: { x: 1 } } });
    const [thread] = await runtime.admin.listThreads();
    expect(thread!.startedWith).toMatchObject({ runId: ran.runId, agent: 'chat', model: 'gpt-4o' });
    expect(thread!.startedWith!.prompt).toBeUndefined();
    expect(thread!.startedWith!.state).toBeUndefined();
    expect(thread!.startedWith!.providerOptions).toBeUndefined();
    const run = (await runtime.admin.getRun(ran.runId!))!.run;
    expect(run.providerOptions ?? null).toBeNull();
  });

  it('records nothing when no level sets provider options', async () => {
    const { runtime } = await makeRuntime({ providerOptions: undefined });
    const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
    const ran = await chat.run({ prompt: 'plain' });
    const [thread] = await runtime.admin.listThreads();
    expect(thread!.startedWith!.providerOptions).toBeNull();
    expect((await runtime.admin.getRun(ran.runId!))!.run.providerOptions).toBeNull();
  });

  it('survives the SQLite store, including its date', async () => {
    const store = new SqliteAdminStore(await openSqlite(':memory:'));
    const { runtime } = await makeRuntime({}, store);
    const chat = runtime.createStreamTextAgent({ name: 'chat', model: 'gpt-4o' });
    const ran = await chat.run({ prompt: 'durable', state: { orgId: 'acme' } });
    // The transition upserts that follow must not clear it
    await store.threads.upsert({ id: ran.threadId, state: 'COMPLETED', model: 'gpt-4o' });
    const [thread] = await store.threads.list({});
    expect(thread!.state).toBe('COMPLETED');
    expect(thread!.startedWith).toMatchObject({ runId: ran.runId, prompt: 'durable', state: { orgId: 'acme' } });
    expect(thread!.startedWith!.at).toBeInstanceOf(Date);
    expect((await store.runs.get(ran.runId!))!.providerOptions).toEqual({ openai: { tier: 'flex' } });
  });
});

File: packages/go-agentenkit/thread_start_test.go
package agentenkit_test

import (
	"testing"

	agentenkit "github.com/eadwinCode/agentic-kit/packages/go-agentenkit"
)

func TestThreadStart_RecordedOnFirstSightWithEveryDispatchParameter(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}), func(c *agentenkit.AgentConfig) {
		c.ProviderOptions = agentenkit.ProviderOptions{"openai": map[string]any{"tier": "flex"}}
	})
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{
		Name: "chat", Model: "gpt-4o", ProviderOptions: agentenkit.ProviderOptions{"anthropic": map[string]any{"effort": "high"}},
	})
	ran := h.run(t, chat, agentenkit.RunInput{
		Prompt: "send the quarterly report", TokenBudget: 5000,
		State:           agentenkit.AgentRunState{"orgId": "acme"},
		ProviderOptions: agentenkit.ProviderOptions{"openai": map[string]any{"tier": "priority"}},
	})
	threads, _ := h.rt.Admin.ListThreads(h.ctx, agentenkit.AdminThreadFilter{})
	sw := threads[0].StartedWith
	if sw == nil {
		t.Fatal("startedWith missing")
	}
	mustEqual(t, sw.RunID, ran.RunID, "runId")
	mustEqual(t, sw.Agent, "chat", "agent")
	mustEqual(t, sw.Model, "gpt-4o", "model")
	mustEqual(t, sw.Prompt, "send the quarterly report", "prompt")
	mustEqual(t, *sw.TokenBudget, 5000, "budget")
	mustEqual(t, sw.State["orgId"], "acme", "state")
	// The three levels merged, the input winning per namespace (§3.1)
	mustEqual(t, sw.ProviderOptions["openai"].(map[string]any)["tier"], "priority", "input wins")
	mustEqual(t, sw.ProviderOptions["anthropic"].(map[string]any)["effort"], "high", "spec namespace kept")
	if sw.At.IsZero() {
		t.Fatal("at")
	}
	mustEqual(t, threads[0].Prompt, "send the quarterly report", "prompt on the summary")
	rec, _ := h.admin.Runs().Get(h.ctx, ran.RunID)
	mustEqual(t, rec.ProviderOptions["openai"].(map[string]any)["tier"], "priority", "run record carries them too")

	// A second run on the thread never overwrites what started it
	h.handleNext(t)
	again := h.run(t, chat, agentenkit.RunInput{ThreadID: ran.ThreadID, Prompt: "and again"})
	threads, _ = h.rt.Admin.ListThreads(h.ctx, agentenkit.AdminThreadFilter{})
	mustEqual(t, threads[0].StartedWith.RunID, ran.RunID, "first run stays")
	mustEqual(t, threads[0].StartedWith.Prompt, "send the quarterly report", "first prompt stays")
	if again.RunID == ran.RunID {
		t.Fatal("second run must have its own id")
	}
	detail, _ := h.rt.Admin.GetThread(h.ctx, ran.ThreadID)
	mustEqual(t, detail.Thread.StartedWith.RunID, ran.RunID, "on the detail too")
}

func TestThreadStart_KeepsOnlyTheIdentityWhenPayloadsAreOff(t *testing.T) {
	h := makeRuntime(t, scripted(step{text: "ok"}), func(c *agentenkit.AgentConfig) { c.RecordPayloads = false })
	chat := h.rt.CreateStreamTextAgent(agentenkit.StreamTextAgentSpec{Name: "chat", Model: "gpt-4o"})
	ran := h.run(t, chat, agentenkit.RunInput{Prompt: "private", State: agentenkit.AgentRunState{"orgId": "acme"},
		ProviderOptions: agentenkit.ProviderOptions{"openai": map[string]any{"x": 1}}})
	threads, _ := h.rt.Admin.ListThreads(h.ctx, agentenkit.AdminThreadFilter{})
	sw := threads[0].StartedWith
	mustEqual(t, sw.RunID, ran.RunID, "runId")
	mustEqual(t, sw.Prompt, "", "no prompt")
	if sw.State != nil || sw.ProviderOptions != nil || sw.TokenBudget != nil {
		t.Fatalf("payloads must be absent: %+v", sw)
	}
	rec, _ := h.admin.Runs().Get(h.ctx, ran.RunID)
	if rec.ProviderOptions != nil {
		t.Fatal("run record must not carry provider options")
	}
}
